README.md
Arduino LoRaWAN library
====================
This repository contains the simple LoRaWAN library originally created by Ideetron B.V. This library is slightly
modified and encapsulated to run in the generic platform, allowing using the SX1272, SX1276 tranceivers and compatible modules (such as some HopeRF RFM9x modules).

This repository is all based on this [repo]( https://git.antares.id/lorawan-loraid/arduino-loraid), with the principal purpose to port it to the frequencies of america US915.

To find out how to use the library itself, see the examples, or see the PDF file in the doc subdirectory.

Features
--------
The arduino LoRaWAN library supports LoRaWAN Class A and Class C implementations operating in EU-868, AS-923 and US-915 bands. Note that this library is fairly simple with the aim of demonstrating the LoRaWAN capabilities.

What certainly works:
 - Sending packets uplink, taking into account duty cycling.
 - Custom frequencies and datarate settings.

What has not been tested:
 - Receiving downlink packets in the RX1 window.
 - Receiving downlink packets in the RX2 window.
 - Receiving and processing MAC commands.
 - Over-the-air activation (OTAA / joining).
 - Class C operation.

If you try one of these untested features and it works, be sure to let
us know (creating a github issue is probably the best way for that).

Installing
----------
To install this library:

 - install it using the Arduino Library manager ("Sketch" -> "Include
   Library" -> "Manage Libraries..."), or
 - download a zipfile from github using the "Download ZIP" button and
   install it using the IDE ("Sketch" -> "Include Library" -> "Add .ZIP
   Library..."
 - clone this git repository into your sketchbook/libraries folder.

For more info, see https://www.arduino.cc/en/Guide/Libraries


Configuration
-------------
A number of features can be configured or disabled by editing the
`config.h` file in the library folder. Unfortunately the Arduino
environment does not offer any way to do this (compile-time)
configuration from the sketch, so be careful to recheck your
configuration when you switch between sketches or update the library.

At the very least, you should set the right type of board in config.h, most other values should be fine at their defaults.

When using the US_915 you need to select which sub-band you will use, by default it is sub-band 6.

Supported hardware
------------------
This library is intended to be used with plain LoRa transceivers,
connecting to them using SPI. In particular, the SX1272 and SX1276
families are supported (which should include SX1273, SX1277, SX1278 and
SX1279 which only differ in the available frequencies, bandwidths and
spreading factors). It has been tested with both SX1272 and SX1276
chips, using the Semtech SX1272 evaluation board and the HopeRF RFM92
and RFM95 boards (which supposedly contain an SX1272 and SX1276 chip
respectively).

Some of the supported pre-built board currently available in the market are:
- [Cytron Shield LoRa-RFM](https://www.cytron.io/p-shield-lora-rfm) 
- [Dragino](http://www.dragino.com/products/module/item/102-lora-shield.html) 
- [Electronic Cats CatWAN USB-Stick](https://www.tindie.com/products/electroniccats/catwan-usb-stick/)
- [Electronic Cats CatWAN Relay Board](https://www.tindie.com/products/electroniccats/catwan-relay-board/)
- [Electronic Cats CatWAN Shield](https://www.tindie.com/products/electroniccats/catwan-shield-lora-and-lorawan-for-arduino/)

This library has been tested using:
- Arduino Uno
- WeMos D1 R2 (ESP8266 family board)

Connections
-----------
To make this library work, your Arduino (or whatever Arduino-compatible
board you are using) should be connected to the transceiver. The exact
connections are a bit dependent on the transceiver board and Arduino
used, so this section tries to explain what each connection is for and
in what cases it is (not) required.


Pin mapping
-----------
As described above, most connections can use arbitrary I/O pins on the
Arduino side. To tell the arduino LoRaWAN library about these, a pin mapping struct is used in the sketch file.

For example, this could look like this:

	sRFM_pins RFM_pins = {
	  	.CS = SS,
	  	.RST = RFM_RST,
	  	.DIO0 = RFM_DIO0,
	  	.DIO1 = RFM_DIO1,
	  	.DIO2 = RFM_DIO2,
	  	.DIO5 = RFM_DIO5
  	}; 
  	
API
--------
This library provides a high-level API for connecting the device to Antares server.

#### LoRa module initialization
Initialize LoRa/LoRaWAN module. Must be called once in the Arduino setup block.

##### Syntax
```c
  void init(void);
```

##### Example
```c
void setup() {
  // Setup loraid access
  lora.init();
  ...
}
```

#### Setup Authentication Keys
Setup authentication keys for your LoRaWAN device, including device address.

##### Syntax
```c
void setAppSKey(unsigned char *ApskKey_in);
void setNwkSKey(unsigned char *NwkKey_in);
```

##### Example
```c
void setup() {
  // Setup loraid access
  if(!lora.init()){
  Serial.println("RFM95 not detected");
  while(1);
  }
  ...

  // Put Antares Key and DevAddress here
  lora.setNwkSKey("b7300d9f68b649ed30530f9dd69f9afe");
  lora.setAppSKey("9d52eef7fab63eda18794d0e503ddf20");
  ...

}
```

#### Setup Device address
Setup device address for activating the device.

##### Syntax
```c
void setDevAddr(unsigned char *devAddr_in);
```

##### Example
```c
void setup() {
  // Setup loraid access
  if(!lora.init()){
  Serial.println("RFM95 not detected");
  while(1);
  }
  ...

  // Put Antares Key and DevAddress here
  lora.setDevAddr("07000007");
  ...

}
```

#### Set Device Class
Set class of the device (Class A or Class C). Input as `CLASS_A` or `CLASS_C` enum.
##### Syntax
```c
void setDeviceClass(devclass_t dev_class);
```

##### Example
```c
void setup() {
  // Setup loraid access
  if(!lora.init()){
  Serial.println("RFM95 not detected");
  while(1);
  }
  ...

  // Set LoRaWAN Class
  lora.setDeviceClass(CLASS_A);
  ...
}

```

#### Set Data Rate
You can set data rate allowed in your region (AS_923, EU_868 or US915).

###	For AS923 or EU868
| data_rate | Name | Config          |Direction
|-----------|------|-----------------|----------------
| 0         | DR0  | SF12 BW 125 KHz | Uplink/Downlink
| 1         | DR1  | SF11 BW 125 KHz | Uplink/Downlink
| 2         | DR2  | SF10 BW 125 KHz | Uplink/Downlink
| 3         | DR3  | SF9 BW 125 KHz  | Uplink/Downlink
| 4         | DR4  | SF8 BW 125 KHz  | Uplink/Downlink
| 5         | DR5  | SF7 BW 125 KHz  | Uplink/Downlink
| 6         | DR6  | SF7 BW 250 KHz  | Uplink/Downlink

### For US915
| data_rate    | Name  | Config          | Direction   
|--------------|-------|-----------------|----------
| 0            | DR0   | SF12 BW 125 KHz | Uplink
| 1            | DR1   | SF11 BW 125 KHz | Uplink
| 2            | DR2   | SF10 BW 125 KHz | Uplink
| 3            | DR3   | SF9 BW 125 KHz  | Uplink
| 4            | DR4   | SF8 BW 500 KHz  | Uplink
| 5:7  	     | RFU   | 		N/A		    | N/A
| 8            | DR8   | SF12 BW 500 KHz  | Downlink
| 9            | DR9   | SF11 BW 500 KHz  | Downlink
| 10           | DR10  | SF10 BW 500 KHz  | Downlink
| 11           | DR11  | SF9 BW 500 KHz  | Downlink
| 12           | DR12  | SF8 BW 500 KHz  | Downlink
| 13           | DR13  | SF7 BW 500 KHz  | Downlink

For US915 is important to remark that DR0-DR4 are only for UPLINKS
and DR8-DR10 are only for DOWNLINKS

*RFU: Reserved for future use

##### Syntax
```c
void setDataRate(unsigned char data_rate);
```

##### Example
```c
void setup() {
  // Setup loraid access
  if(!lora.init()){
  Serial.println("RFM95 not detected");
  while(1);
  }
  ...

  // Set Data Rate to SF10 BW 125 KHz

doc/release-notes.txt
==============================================================================
Arduino LoRa.id SDK v1.0.0 (2-May-2018)
-------------------------------

 - Support Class A and Class C

 - Tested using Arduino Uno and LoRa RFM modules

 - Only send LoRa RX to Serial

==============================================================================

src/arduino-rfm/RFM95.cpp
/******************************************************************************************
* Copyright 2017 Ideetron B.V.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
******************************************************************************************/
/****************************************************************************************
* File:     RFM95.cpp
* Author:   Gerben den Hartog
* Compagny: Ideetron B.V.
* Website:  http://www.ideetron.nl/LoRa
* E-mail:   info@ideetron.nl
****************************************************************************************/
/****************************************************************************************
* Created on:         06-01-2017
* Supported Hardware: ID150119-02 Nexus board with RFM95
****************************************************************************************/

#include <SPI.h>
#include "LoRaMAC.h"
#include "RFM95.h"
#include "lorawan-arduino-rfm.h"
#include <Arduino.h>
#include "Struct.h"
#include "Commands.h"
#include "Config.h"

#ifdef US_915
  #ifdef SUBND_1
  const unsigned char LoRa_Frequency[9][3] = {
	{ 0xE1, 0xF9, 0xC0 },		//Channel 0 903.900 MHz / 61.035 Hz = 14809536 = 0xE1F9C0
	{ 0xE2, 0x06, 0x8C },		//Channel 1 904.100 MHz / 61.035 Hz = 14812812 = 0xE2068C
	{ 0xE2, 0x13, 0x59 },		//Channel 2 904.300 MHz / 61.035 Hz = 14816089 = 0xE21359
	{ 0xE2, 0x20, 0x26 },		//Channel 3 904.500 MHz / 61.035 Hz = 14819366 = 0xE22026
	{ 0xE2, 0x2C, 0xF3 },		//Channel 4 904.700 MHz / 61.035 Hz = 14822643 = 0xE22CF3
	{ 0xE2, 0x39, 0xC0 },		//Channel 5 904.900 MHz / 61.035 Hz = 14825920 = 0xE239C0
	{ 0xE2, 0x46, 0x8C },		//Channel 6 905.100 MHz / 61.035 Hz = 14829196 = 0xE2468C
	{ 0xE2, 0x53, 0x59 },		//Channel 7 905.300 MHz / 61.035 Hz = 14832473 = 0xE25359
  { 0xE6, 0xD3, 0x5A },   //Channel RX2 window 923.3 MHz / 61.035 Hz = 0xE6D35A
  };  
  #else
  const unsigned char LoRa_Frequency[9][3] = {
    { 0xE3, 0xD3, 0x59 }, //Channel 48 911.900 MHz / 61.035 Hz = 0xE3D359
    { 0xE4, 0x06, 0x8C }, //Channel 49 912.100 MHz / 61.035 Hz = 0xE4068C
    { 0xE4, 0x13, 0x59 }, //Channel 50 912.300 MHz / 61.035 Hz = 0xE41359
    { 0xE4, 0x20, 0x26 }, //Channel 51 912.500 MHz / 61.035 Hz = 0xE42026
    { 0xE4, 0x2C, 0xF3 }, //Channel 52 912.700 MHz / 61.035 Hz = 0xE42CF3
    { 0xE4, 0x39, 0xBF }, //Channel 53 912.900 MHz / 61.035 Hz = 0xE439BF
    { 0xE4, 0x46, 0x8C }, //Channel 54 913.100 MHz / 61.035 Hz = 0xE4468C
    { 0xE4, 0x53, 0x59 }, //Channel 55 913.300 MHz / 61.035 Hz = 0xE45359
    { 0xE6, 0xD3, 0x5A }, //Channel RX2 window 923.3 MHz / 61.035 Hz = 0xE6D35A
  };
  #endif
#endif
/*
*****************************************************************************************
* Description: Function used to initialize the RFM module on startup
*****************************************************************************************
*/

bool RFM_Init()
{
  
  uint8_t ver = RFM_Read(0x42);
  if(ver!=18){
    return 0;
  }
  //Switch RFM to sleep
  //DONT USE Switch mode function
  RFM_Write(0x01,0x00);

  //Wait until RFM is in sleep mode
  delay(50);
 
  //Set RFM in LoRa mode
  //DONT USE Switch mode function
  RFM_Write(0x01,0x80);
  //Swtich RFM to standby
  RFM_Switch_Mode(0x01);
  //Set channel to channel 0
  RFM_Change_Channel(0x00);
  //PA pin (minimal power)
  //RFM_Write(0x09,0xF0);
  //set to 17dbm
  RFM_Write(0x09,0xF0);


  //Switch LNA boost on
  RFM_Write(0x0C,0x23);

  //Set RFM To datarate 0 SF12 BW 125 kHz
  RFM_Change_Datarate(0x00);

  //Rx Timeout set to 37 symbols
  RFM_Write(0x1F,0x25);

  //Preamble length set to 8 symbols
  //0x0008 + 4 = 12
  RFM_Write(0x20,0x00);
  RFM_Write(0x21,0x08);

  //Set LoRa sync word
  RFM_Write(0x39,0x34);

  //Set FIFO pointers
  //TX base adress
  RFM_Write(0x0E,0x80);
  //Rx base adress
  RFM_Write(0x0F,0x00);
  return 1;
}

/*
*****************************************************************************************
* Description : Function for sending a package with the RFM
*
* Arguments   : *RFM_Tx_Package pointer to buffer with data and counter of data
*               *LoRa_Settings pointer to sSettings struct
*****************************************************************************************
*/

void RFM_Send_Package(sBuffer *RFM_Tx_Package, sSettings *LoRa_Settings)
{
  unsigned char i;
  unsigned char RFM_Tx_Location = 0x00;

  //Set RFM in Standby mode
  RFM_Switch_Mode(0x01);

  //Switch Datarate
  RFM_Change_Datarate(LoRa_Settings->Datarate_Tx);

  //Switch Channel
  RFM_Change_Channel(LoRa_Settings->Channel_Tx);

  //Switch DIO0 to TxDone
  RFM_Write(0x40,0x40);

  //Set IQ to normal values
  RFM_Write(0x33,0x27);
  RFM_Write(0x3B,0x1D);

  //Set payload length to the right length
  RFM_Write(0x22,RFM_Tx_Package->Counter);

  //Get location of Tx part of FiFo
  RFM_Tx_Location = RFM_Read(0x0E);

  //Set SPI pointer to start of Tx part in FiFo
  RFM_Write(0x0D,RFM_Tx_Location);

  //Write Payload to FiFo
  for (i = 0;i < RFM_Tx_Package->Counter; i++)
  {
    RFM_Write(0x00, RFM_Tx_Package->Data[i]);
  }

  //Switch RFM to Tx
  RFM_Write(0x01,0x83);

  //Wait for TxDone
  while(digitalRead(RFM_pins.DIO0) == LOW)
  {
  }

  //Clear interrupt
  RFM_Write(0x12,0x08);

  //Switch RFM back to receive if it is a type C mote
  if(LoRa_Settings->Mote_Class == 0x01)
  {
    //Switch Back to Continuous receive
    RFM_Continuous_Receive(LoRa_Settings);
  }
}

/*
*****************************************************************************************
* Description : Function to switch RFM to single receive mode, used for Class A motes
*
* Arguments   : *LoRa_Settings pointer to sSettings struct
*
* Return	  : message_t Status of the received message
*****************************************************************************************
*/

message_t RFM_Single_Receive(sSettings *LoRa_Settings)
{
  message_t Message_Status = NO_MESSAGE;

  //Change DIO 0 back to RxDone
  RFM_Write(0x40,0x00);

  //Invert IQ Back
  RFM_Write(0x33,0x67);
  RFM_Write(0x3B,0x19);

  //Change Datarate
  RFM_Change_Datarate(LoRa_Settings->Datarate_Rx);

  //Change Channel
  RFM_Change_Channel(LoRa_Settings->Channel_Rx);

  //Switch RFM to Single reception
  RFM_Switch_Mode(0x06);

  //Wait until RxDone or Timeout
  //Wait until timeout or RxDone interrupt
  while((digitalRead(RFM_pins.DIO0) == LOW) && (digitalRead(RFM_pins.DIO1) == LOW))
  {
  }

  //Check for Timeout
  if(digitalRead(RFM_pins.DIO1) == HIGH)
  {
    //Clear interrupt register
    RFM_Write(0x12,0xE0);
    Message_Status = TIMEOUT;
    UART_Send_Newline();
    UART_Send_Newline();
  }

  //Check for RxDone
  if(digitalRead(RFM_pins.DIO0) == HIGH)
  {
	  Message_Status = NEW_MESSAGE;
  }

  return Message_Status;
}


/*
*****************************************************************************************
* Description : Function to switch RFM to continuous receive mode, used for Class C motes
*
* Arguments   : *LoRa_Settings pointer to sSettings struct
*****************************************************************************************
*/
void RFM_Continuous_Receive(sSettings *LoRa_Settings)
{
  //Change DIO 0 back to RxDone
  RFM_Write(0x40,0x00);

  //Invert IQ Back
  RFM_Write(0x33,0x67);
  RFM_Write(0x3B,0x19);
  
	//Change Datarate
	RFM_Change_Datarate(LoRa_Settings->Datarate_Rx);

	//Change Channel
	RFM_Change_Channel(LoRa_Settings->Channel_Rx);

	//Switch to continuous receive
	RFM_Switch_Mode(0x05);

}

/*
*****************************************************************************************
* Description : Function to retrieve a message received by the RFM
*
* Arguments   : *RFM_Rx_Package pointer to sBuffer struct containing the data received
*				and number of bytes received
*
* Return	  : message_t Status of the received message
*****************************************************************************************
*/

message_t RFM_Get_Package(sBuffer *RFM_Rx_Package)
{
  unsigned char i;
  unsigned char RFM_Interrupts = 0x00;
  unsigned char RFM_Package_Location = 0x00;
  message_t Message_Status;

  //Get interrupt register
  RFM_Interrupts = RFM_Read(0x12);

  //UART_Send_Data(RFM_Interrupts,0x01);

  //Check CRC
  if((RFM_Interrupts & 0x20) != 0x20)
  {
	  Message_Status = CRC_OK;

	  Serial.write("CRC OK");

	  UART_Send_Newline();
  }
  else
  {
	  Message_Status = WRONG_MESSAGE;

	  Serial.write("CRC NOK");

	  UART_Send_Newline();
  }

  RFM_Package_Location = RFM_Read(0x10); /*Read start position of received package*/
  RFM_Rx_Package->Counter = RFM_Read(0x13); /*Read length of received package*/

  RFM_Write(0x0D,RFM_Package_Location); /*Set SPI pointer to start of package*/

  for (i = 0x00; i < RFM_Rx_Package->Counter; i++)
  {
    RFM_Rx_Package->Data[i] = RFM_Read(0x00);
  }

  //Clear interrupt register
  RFM_Write(0x12,0xE0);

  return Message_Status;
}

/*
*****************************************************************************************
* Description : Funtion that reads a register from the RFM and returns the value
*
* Arguments   : RFM_Address Address of register to be read
*
* Returns   : Value of the register
*****************************************************************************************
*/

unsigned char RFM_Read(unsigned char RFM_Address)
{
  unsigned char RFM_Data;

  //Set NSS pin low to start SPI communication
  digitalWrite(RFM_pins.CS,LOW);

  //Send Address
  SPI.transfer(RFM_Address);
  //Send 0x00 to be able to receive the answer from the RFM
  RFM_Data = SPI.transfer(0x00);

  //Set NSS high to end communication
  digitalWrite(RFM_pins.CS,HIGH);

  //Return received data
  return RFM_Data;
}

/*
*****************************************************************************************
* Description : Funtion that writes a register from the RFM
*
* Arguments   : RFM_Address Address of register to be written
*         RFM_Data    Data to be written
*****************************************************************************************
*/

void RFM_Write(unsigned char RFM_Address, unsigned char RFM_Data)
{
  //Set NSS pin Low to start communication
  digitalWrite(RFM_pins.CS,LOW);

  //Send Addres with MSB 1 to make it a writ command
  SPI.transfer(RFM_Address | 0x80);
  //Send Data
  SPI.transfer(RFM_Data);

  //Set NSS pin High to end communication
  digitalWrite(RFM_pins.CS,HIGH);
}

/*
*****************************************************************************************
* Description : Function to change the datarate of the RFM module. Setting the following
*				register: Spreading factor, Bandwith and low datarate optimisation.
*
* Arguments   : Datarate the datarate to set
*****************************************************************************************
*/
void RFM_Change_Datarate(unsigned char Datarate)
{
	switch(Datarate)
	{
		#ifndef US_915
    case 0x00: //SF12 BW 125 kHz
			RFM_Write(0x1E,0xC4); //SF12 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x0C); //Low datarate optimization on AGC auto on
			break;
		case 0x01: //SF11 BW 125 kHz
			RFM_Write(0x1E,0xB4); //SF11 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x0C); //Low datarate optimization on AGC auto on
			break;
		case 0x02: //SF10 BW 125 kHz
			RFM_Write(0x1E,0xA4); //SF10 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
			break;
		case 0x03: //SF9 BW 125 kHz
			RFM_Write(0x1E,0x94); //SF9 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
			break;
		case 0x04: //SF8 BW 125 kHz
			RFM_Write(0x1E,0x84); //SF8 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
			break;
		case 0x05: //SF7 BW 125 kHz
			RFM_Write(0x1E,0x74); //SF7 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    case 0x06: //SF7 BW 250kHz
			RFM_Write(0x1E,0x74); //SF7 CRC On
			RFM_Write(0x1D,0x82); //250 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    // Its defined US915 cause the datarates are differents 
    #else 
    case 0x00: //SF10 BW 125 kHz
			RFM_Write(0x1E,0xA4); //SF10 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x0C); //Low datarate optimization on AGC auto on
			break;
		case 0x01: //SF9 BW 125 kHz
			RFM_Write(0x1E,0x94); //SF9 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x0C); //Low datarate optimization on AGC auto on
			break;
		case 0x02: //SF8 BW 125 kHz
			RFM_Write(0x1E,0x84); //SF8 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
			break;
		case 0x03: //SF7 BW 125 kHz
			RFM_Write(0x1E,0x74); //SF7 CRC On
			RFM_Write(0x1D,0x72); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
			break;
		case 0x04: //SF8 BW 500 kHz
			RFM_Write(0x1E,0x84); //SF8 CRC On
			RFM_Write(0x1D,0x92); //500 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
			break;
		case 0x08: //SF12 BW 500 kHz
			RFM_Write(0x1E,0xC4); //SF12 CRC On
			RFM_Write(0x1D,0x92); //125 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    case 0x09: //SF11 BW 500kHz
			RFM_Write(0x1E,0xB4); //SF11 CRC On
			RFM_Write(0x1D,0x92); //500 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    case 0x0A: //SF10 BW 500kHz
			RFM_Write(0x1E,0xA4); //SF10 CRC On
			RFM_Write(0x1D,0x92); //500 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    case 0x0B: //SF9 BW 500kHz
			RFM_Write(0x1E,0x94); //SF9 CRC On
			RFM_Write(0x1D,0x92); //500 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    case 0x0C: //SF8 BW 500kHz
			RFM_Write(0x1E,0x84); //SF8 CRC On
			RFM_Write(0x1D,0x92); //500 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    case 0x0D: //SF7 BW 500kHz
			RFM_Write(0x1E,0x74); //SF7 CRC On
			RFM_Write(0x1D,0x92); //500 kHz 4/5 coding rate explicit header mode
			RFM_Write(0x26,0x04); //Low datarate optimization off AGC auto on
      break;
    #endif
	}
}

/*
*****************************************************************************************
* Description : Function to change the channel of the RFM module. Setting the following
*				register: Channel
*
* Arguments   : Channel the channel to set
*****************************************************************************************
*/
void RFM_Change_Channel(unsigned char Channel)
{
	switch(Channel)
	{ 
    //AS_923 Frequency select
    #if defined(AS_923)
		case 0x00: //Channel 0 923.200 MHz / 61.035 Hz = 15125748 = 0xE6CCF3
    case 0x10: //Receive channel 923.200 MHz / 61.035 Hz = 15125748 = 0xE6CCF3
			RFM_Write(0x06,0xE6);
			RFM_Write(0x07,0xCC);
			RFM_Write(0x08,0xF3);
      break;
    case 0x01: //Channel 1 923.400 MHz / 61.035 Hz = 15129024 = 0xE6D9C0
			RFM_Write(0x06,0xE6);
			RFM_Write(0x07,0xD9);
			RFM_Write(0x08,0xC0);
      break;
		case 0x02: //Channel 2 923.600 MHz / 61.035 Hz = 15132301 = 0xE6E68D
			RFM_Write(0x06,0xE6);
			RFM_Write(0x07,0xE6);
			RFM_Write(0x08,0x8D);
      break;
		case 0x03: //Channel 3 923.800 MHz / 61.035 Hz = 15135578 = 0xE6F35A
			RFM_Write(0x06,0xE6);
			RFM_Write(0x07,0xF3);
			RFM_Write(0x08,0x5A);
      break;
		case 0x04: //Channel 4 924.000 MHz / 61.035 Hz = 15138855 = 0xE70027
			RFM_Write(0x06,0xE7);
			RFM_Write(0x07,0x00);
			RFM_Write(0x08,0x27);
      break;
		case 0x05: //Channel 5 924.200 MHz / 61.035 Hz = 15142132 = 0xE70CF4
			RFM_Write(0x06,0xE7);
			RFM_Write(0x07,0x0C);
			RFM_Write(0x08,0xF4);
      break;
		case 0x06: //Channel 6 924.400 MHz / 61.035 Hz = 15145408 = 0xE719C0
			RFM_Write(0x06,0xE7);
			RFM_Write(0x07,0x19);
			RFM_Write(0x08,0xC0);
      break;
		case 0x07: //Channel 7 924.600 MHz / 61.035 Hz = 15148685 = 0xE7268D
			RFM_Write(0x06,0xE7);
			RFM_Write(0x07,0x26);
			RFM_Write(0x08,0x8D);
      break;
		case 0x08: //Channel 8 924.400 MHz / 61.035 Hz = 15151962 = 0xE7335A
			RFM_Write(0x06,0xE7);
			RFM_Write(0x07,0x33);
			RFM_Write(0x08,0x5A);
      break;
    //EU_868 Frequency select
    #elif defined(EU_868)
 		case 0x00: //Channel 0 868.100 MHz / 61.035 Hz = 14222987 = 0xD9068B
			RFM_Write(0x06,0xD9);
			RFM_Write(0x07,0x06);
			RFM_Write(0x08,0x8B);
			break;
		case 0x01: //Channel 1 868.300 MHz / 61.035 Hz = 14226264 = 0xD91358
			RFM_Write(0x06,0xD9);
			RFM_Write(0x07,0x13);
			RFM_Write(0x08,0x58);
			break;
		case 0x02: //Channel 2 868.500 MHz / 61.035 Hz = 14229540 = 0xD92024
			RFM_Write(0x06,0xD9);
			RFM_Write(0x07,0x20);
			RFM_Write(0x08,0x24);
			break;
		case 0x03: //Channel 3 867.100 MHz / 61.035 Hz = 14206603 = 0xD8C68B
			RFM_Write(0x06,0xD8);
			RFM_Write(0x07,0xC6);
			RFM_Write(0x08,0x8B);
			break;
		case 0x04: //Channel 4 867.300 MHz / 61.035 Hz = 14209880 = 0xD8D358
			RFM_Write(0x06,0xD8);
			RFM_Write(0x07,0xD3);
			RFM_Write(0x08,0x58);
			break;
		case 0x05: //Channel 5 867.500 MHz / 61.035 Hz = 14213156 = 0xD8E024
			RFM_Write(0x06,0xD8);
			RFM_Write(0x07,0xE0);
			RFM_Write(0x08,0x24);
			break;
		case 0x06: //Channel 6 867.700 MHz / 61.035 Hz = 14216433 = 0xD8ECF1
			RFM_Write(0x06,0xD8);
			RFM_Write(0x07,0xEC);
			RFM_Write(0x08,0xF1);
			break;
		case 0x07: //Channel 7 867.900 MHz / 61.035 Hz = 14219710 = 0xD8F9BE
			RFM_Write(0x06,0xD8);
			RFM_Write(0x07,0xF9);
			RFM_Write(0x08,0xBE);
			break;
		case 0x10: //Receive channel 869.525 MHz / 61.035 Hz = 14246334 = 0xD961BE
			RFM_Write(0x06,0xD9);
			RFM_Write(0x07,0x61);
			RFM_Write(0x08,0xBE);
			break;
    //US_915 Frequency select
    #else
    case 0x00: //Channel 0 911.900 MHz / 61.035 Hz = 0xE3D359
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[0][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[0][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[0][2])));
			break;
		case 0x01: //Channel 1 912.100 MHz / 61.035 Hz = 0xE4068C
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[1][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[1][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[1][2])));
			break;
		case 0x02: //Channel 2 912.300 MHz / 61.035 Hz = 0xE41359   
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[2][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[2][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[2][2])));
			break;
		case 0x03: //Channel 3 912.500 MHz / 61.035 Hz = 0xE42026
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[3][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[3][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[3][2])));
			break;
		case 0x04: //Channel 4 912.700 MHz / 61.035 Hz = 0xE42CF3
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[4][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[4][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[4][2])));
			break;
		case 0x05: //Channel 5 912.900 MHz / 61.035 Hz = 0xE439BF
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[5][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[5][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[5][2])));
			break;
		case 0x06: //Channel 6 913.100 MHz / 61.035 Hz = 0xE4468C
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[6][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[6][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[6][2])));
			break;
		case 0x07: //Channel 7 912.300 MHz / 61.035 Hz = 0xE45359
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[7][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[7][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[7][2])));
			break;
		case 0x10: //Receive channel 923.3 MHz / 61.035 Hz = 0xE6D35A
			RFM_Write(0x06,pgm_read_byte(&(LoRa_Frequency[8][0])));
			RFM_Write(0x07,pgm_read_byte(&(LoRa_Frequency[8][1])));
			RFM_Write(0x08,pgm_read_byte(&(LoRa_Frequency[8][2])));
			break;
    #endif
	}
}

/*
*****************************************************************************************
* Description : Function to change the operation mode of the RFM. Switching mode and wait
*				for mode ready flag
*               DO NOT USE FOR SLEEP
*
* Arguments   : Mode the mode to set
*****************************************************************************************
*/
void RFM_Switch_Mode(unsigned char Mode)
{
    Mode = Mode | 0x80; //Set high bit for LoRa mode

    //Switch mode on RFM module
    RFM_Write(0x01,Mode);

	//Wait on mode ready
  #ifdef BOARD_DRAGINO_SHIELD
    // while(digitalRead(RFM_pins.DIO5) == LOW)
    // {
    // }
  #endif
}

